doc/compiler/optimizer/ValuePropagation.md

```diff
@@ -0,0 +1,191 @@
+
+Note that two nodes may represent the same value need not always have the same value number. However, if two nodes represent different values then their value numbers _must_ also be different.
+
+Value numbering encompasses notions like definitions and uses. For example, if we take the value `4` and store it to a temporary, say `y = 4`, and load this temporary value several blocks later. The value numbering analysis is capable of realizing that it is the same value flowing through all those areas and it will give the same value number to the right hand side of the store and to the left hand side of the store, in addition to all loads that are fed uniquely by that store.
+
+This is the simplest example when value numbers kind of flow through and are the same, and give us a sense that we are talking about the same value which has the same underling properties.
+
+On the contrary, to give us an intuitive sense for a case when we won't have the same value number is to examine what happens during control flow:
+
+```
+if (...) {
+    x = 1;
+} else {
+    x = 2;
+}
+
+y = x;
+```
+
+In this case the value number of the right hand side of the assignment `y = x` will be different than the value number of the left hand side of the stores to `x` in the if-then-else blocks, because at compile time we may not be able to determine which path we would take, so all three nodes will have different value numbers.
+
+Another important property of value numbers that we should understand is that the notion is beyond just syntax. We can have two expressions such as `x + 5` and `y + 5` where `x` and `y` are known to be equal over a particular range, and thus the two addition expressions will also have the same value number. Thus value numbers are not all just trivial use-def chains and parent-child sort of relations. We can have different nodes (expressions) that don't particularly look alike have the same value number.
+
+This is a very high level summary of what value numbering is. There are excellent resources in literature or online to learn more about value numbering.
+
+## Constraints and Handlers
+
+[Value propagation constraints](https://github.com/eclipse/omr/blob/25f479d2b671c2a0a03156cc0a48024a46d2e3c2/compiler/optimizer/VPConstraint.hpp) are the fundamental classes which are flowed around by the value propagation analysis. Their purpose is the represent constraints determined by the analysis on _value numbers_. If for example we had an expression such as `x = 5` then we would create a constraint which represents the constant value of `5` and we would associate this constraint with the value numbers for the nodes representing the left hand side (`iconst`) and right hand side (`istore`) nodes. Remember that these two nodes would have the same value number because we can prove that they are equal. The VP handlers consume these relations between value numbers and constraints to optimize the program.
```

Comment:
`left hand side (`iconst`) and right hand side (`istore`) nodes` should probably be the other way around

Reply:
Fixed in [force-push](https://github.com/eclipse/omr/compare/76b1afbc94d81d8c76ee613e2e89e6c3aa4aab43..9fb1b0b92e8935d254254edca4bd50d3a6c93e48).